### PR TITLE
utime module documentation fixes and cleanup

### PR DESCRIPTION
**Pull request by SpotlightKid**
* Fix mis-spelling of `ticks_add` in code examples.
* Be consistent about parentheses after function names.
* Be consistent about formatting of function, variable and constant names.
* Be consistent about spaces and punctuation.
* Fix some language errors (missing or wrong words, wrong word order).
* Keep line length under 90 chars.

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

### CODE REVIEW

File: docs/library/utime.rst [hunk 1]
```diff
@@ -28,8 +28,8 @@ by following means:
   in such case).
 
 If actual calendar time is not maintained with a system/MicroPython RTC,
-functions below which require reference to current absolute time may
-behave not as expected.
+functions below which require reference to current absolute time may not
```
Comment (pfalcon) [inline docs/library/utime.rst]: You're of course using "may not" in the sense "not allowed to", as in "You may not make changes for bad" ? If so, your change is incorrect. And if you don't use "may not" in this sense, then why do you make a change which replaces the unambiguous phrase with an ambiguous one, which will cause confusion?

Comment (SpotlightKid) [inline docs/library/utime.rst]: No, I'm using it in the sense of "will possibly not". "may behave not" is just wrong word order in English.

Comment (pfalcon) [inline docs/library/utime.rst]: > No, I'm using it in the sense of "will possibly not".

Then people don't understand you and get confused. Because you probably then mean the same as in "value of N % 4 may not reach a value of 4 for any N".

> "may behave not" is just wrong word order in English.

Said who? (Another "wrong" word order.)

Comment (SpotlightKid) [inline docs/library/utime.rst]: Perhaps "might not" then?

Comment (pfalcon) [inline docs/library/utime.rst]: I can assure you that formulations like above were purposedly (purposely?) crafted in a way to minimize ambiguity based on the formal structure of English as a *language* (in a formal sense, even if it's still a highly ambiguous natural language).

The most obvious (almost the only possible) way to parse that phrase is:

> functions [...] may (behave (not (as expected)))

That means that "not" has no direct syntactical relation to "may", as indeed, English phrase "may not" is highly ambiguous (same applies to its past tense, thought indeed, connotations are different). If you want a second opinion on that refer to the list of formally defined modals from https://www.ietf.org/rfc/rfc2119.txt . While it includes "must" vs "must not" and "should" vs "should not", it includes only "may". The reason for that is what we discussed above - any phrase with "may not" is inherently ambiguous, so any doc faithful to RFC2119 would never put those 2 words in a row. And MicroPython docs try to follow these best practices.

The bottom line which would be nice for any contributor to keep in mind: many things in MicroPython are done for a reason. Not all, like there's no good reason in typos and missing articles, yet many things definitely have a reason behind them. Telling one from another may be not easy, but possible (and then it's better to err on fixing the obvious simple things rather than "everything at once").

Comment (SpotlightKid) [inline docs/library/utime.rst]: Just tell me what to do with the PR.

Comment (pfalcon) [inline docs/library/utime.rst]: Yes, I'd suggest to undo this specific change, then review of the rest of changes can continue (I did just a quick look thru them, and they look good, but I apologize in advance if there will be more change suggestions). I'm just trying to explain an idea behind why it was written like that.

Comment (SpotlightKid) [inline docs/library/utime.rst]: I just pushed a commit that reverts this particular change.

**Comment by pfalcon** [comment]
Sorry for the delay with further review - I was travelling at that time and now drowned in real-world things. In the meantime, you could squash the commits together...

**Comment by SpotlightKid** [comment]
Rebased on master and squashed.

**Comment by pfalcon** [comment]
Merged, thanks!

> Be consistent about parentheses after function names.

For my own reference, that's how CPython docs have it, so it's definitely right.

> Keep line length under 90 chars.

Note that going and reformatting lines to be under some length isn't a good idea on its own (mostly adds noise to git history/annotation). However, if there's a bigger reason to reformat something (like an addition causes a lot of lines to reflow), I'd saying trying to keep it under 75 chars is even better idea than 90 (but then it still will cause noisy patches, which is harder (== longer) to review).

Thanks again.

**Comment by SpotlightKid** [comment]
Yes, I admit choosing the 90 char line length limit was a bit arbitrary, but that was what most of the paragraphs seemed to follow, except for a few outliers.

I think it would be good to set a few formatting guidelines for the documentation. Maybe those could be added to https://github.com/micropython/micropython/wiki/ContributorGuidelines or a sub-page thereof. Or to the README in the `docs` directory. What do you think?

**Comment by pfalcon** [comment]
In principle I think it's good, but I'm not sure we're ready for this, like, there're way too many things to do on docs, and things like line length are relatively minor to take care of, so I'm not sure we'd have critical mass of people who'd care to settle that now (or soon). All in all, any specific line length limit would need to be approved by @dpgeorge.

**Comment by SpotlightKid** [comment]
IMHO, the documentation urgently needs some love and the longer this is neglected, the worse it will get. I have a few PRs in the pipeline for fixes to other library documentation pages and while working on them I constantly encounter places in the documentation, where I think, "if only there had been some (written) guidelines in place, maybe this wouldn't be such a mess".

I'm not talking only about line-length, but also:

* Consistent punctuation
* Correct ReST syntax (e.g. double instead single backticks, missing double colons, etc.)
* Consistent use of vanilla ReST markup (e.g. warning directives) and Sphinx references
* Consistently correct grammar (e.g. use of articles)
* Consistent formatting of code identifiers (backticks or not, parentheses after function/method names or not, etc., whitespace e.g. parameter lists)
* PEP-8 compliance in code examples
* PEP-257 compliance in function/method docs

And of course there also should be some agreement on how much documentation there needs to be and how to structure it (on a page level) , e.g. how to document differences between the ports, etc. At the moment this is handled in some places but not consistently everywhere.

I know that this is potentially a big task, but it is also something that belongs into the "Clearing the backlog" (#2602) category, as @dpgeorge has written himself (item 3 there).

Would it help if I opened a ticket about creating "Documentation Guidelines" where we could collect these things and (hopefully) come to an initial set of guidelines (which can always be amended later)?

**Comment by pfalcon** [comment]
> IMHO, the documentation urgently needs some love and the longer this is neglected, the worse it will get.

I'm glad you noticed ;-). Here's my ticket which summarizes my doubts: https://github.com/micropython/micropython/issues/2578 , as you can see it's 5 months old, and I issued "btw warnings" similar to yours before and after that, so again, I'm glad there's someone else seeing it as such. But there's nothing "urgent", in the sense that there's nothing which has to be done today which should not have been done months ago. And things are being done, albeit much more slowly than I'd want. You're welcome to join effort on *sustainable* improvements of docs.

 > Would it help if I opened a ticket about creating "Documentation Guidelines" where we could collect these things and (hopefully) come to an initial set of guidelines (which can always be amended later)?

It wouldn't hurt, but I don't think that it's the most effective way to resolve the situation either. What I would consider efficient sustainable way of improvement is: select one point from your list above, and resolve that single point completely. This one:

> Correct ReST syntax (e.g. double instead single backticks

sounds good for starters. The patch fixing that will be easy to review and merge. Well, I may imagine that producing single patch for entire doc may be still cumbersome, so feel free to do that in ~3 patches, submitted one by one, merge of previous one fueling your work on the next. When you resolve 3-4 simple points like that, it may become a good time to draft the documentation guidelines you talk about.

In either case, it's just my opinion, feel free to do it the way you see fit, but please remember: sustainability is the key. I'd estimate the effort you're talking about to take some 3 months, so please make sure you're ready to spend those 3 months (and that's why I recommend to instead to start with mundane consistent fixes - the immediate help to the project and no premature commitment from your side).
